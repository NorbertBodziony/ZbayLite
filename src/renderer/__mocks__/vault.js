export default () => ({
  create: jest.fn(async () => null),
  unlock: jest.fn(async () => null),
  exists: jest.fn(() => false)
})
