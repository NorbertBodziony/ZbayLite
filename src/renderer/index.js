import { Web } from './vendor/buttercup'
import React from 'react'
import { render } from 'react-dom'
import { ipcRenderer } from 'electron'

import Root from './Root'
import store from './store'
import nodeHandlers from './store/handlers/node'
import updateHandlers from './store/handlers/update'

Web.HashingTools.patchCorePBKDF()

ipcRenderer.on('bootstrappingNode', (event, { bootstrapping, message }) => {
  store.dispatch(nodeHandlers.actions.setBootstrapping(bootstrapping))
  store.dispatch(nodeHandlers.actions.setBootstrappingMessage(message))
})

ipcRenderer.on('newUpdateAvailable', (event) => {
  store.dispatch(updateHandlers.epics.checkForUpdate())
})

window.jdenticon_config = {
  lightness: {
    color: [0.31, 0.44],
    grayscale: [0.52, 0.57]
  },
  saturation: {
    color: 0.82,
    grayscale: 0.84
  },
  backColor: '#f3f0f6ff'
}

render(<Root />, document.getElementById('root'))

module.hot.accept()
