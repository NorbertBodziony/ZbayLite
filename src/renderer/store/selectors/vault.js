import { createSelector } from 'reselect'

const store = s => s

const vault = createSelector(store, state => state.get('vault'))

const exists = createSelector(vault, v => v.exists)
const creating = createSelector(vault, v => v.creating)
// TODO: 14/05 this should be checked on vault, not in redux
const locked = createSelector(vault, v => v.locked)
const unlocking = createSelector(vault, v => v.unlocking)
const error = createSelector(vault, v => v.error)

export default {
  exists,
  creating,
  unlocking,
  error,
  locked
}
