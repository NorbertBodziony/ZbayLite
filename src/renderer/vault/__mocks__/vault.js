import channelsFactory from '../channels'

const workspace = jest.mock()
workspace.save = jest.fn()

export const mock = {
  workspace,
  setArchive: (archive) => {
    workspace.archive = archive
  }
}

export default class Vault {
  constructor (sourceCredentials, archiveCredentials) {
    this.channels = channelsFactory(this)
    this.withWorkspace = async (cb) => cb(workspace)
    this.lock = jest.fn()
  }
}
