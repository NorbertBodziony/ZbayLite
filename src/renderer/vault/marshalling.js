import { Archive, Workspace, Credentials, Datasources } from '../vendor/buttercup'
import { validateArchiveGroups } from './validateArchive'

export const archiveGroups = ['Identities', 'Channels', 'Contacts', 'Offers', 'TxnTs', 'Adverts', 'RemovedChannels']

export const createArchive = () => {
  const archive = new Archive()
  archiveGroups.forEach(group => {
    archive.createGroup(group)
  })

  return archive
}

export const credentialsToDatasource = ({ sourceCredentials }) => {
  const datasourceDescriptionRaw = sourceCredentials.getValueOrFail('datasource')
  const datasourceDescription =
    typeof datasourceDescriptionRaw === 'string'
      ? JSON.parse(datasourceDescriptionRaw)
      : datasourceDescriptionRaw

  if (typeof datasourceDescription.type !== 'string') {
    throw new Error('Invalid or missing type')
  }
  return Datasources.objectToDatasource(datasourceDescription, sourceCredentials)
}

export const credentialsToWorkspace = async ({
  sourceCredentials,
  archiveCredentials,
  createSource = false
}) => {
  const datasource = credentialsToDatasource({ sourceCredentials })
  let archive = null
  if (createSource) {
    const defaultArchive = createArchive()
    await datasource.save(defaultArchive.getHistory(), archiveCredentials)
    archive = defaultArchive
  } else {
    const history = await datasource.load(archiveCredentials)
    archive = Archive.createFromHistory(history)
    validateArchiveGroups(archive)
  }
  const workspace = new Workspace()
  workspace.setArchive(archive, datasource, archiveCredentials)
  return workspace
}

export const passwordToSecureStrings = async ({ masterPassword, datasourceObj }) => {
  const passwordCredentials = Credentials.fromPassword(masterPassword)
  const sourceCredentials = new Credentials({
    ...passwordCredentials,
    type: datasourceObj.type
  })
  sourceCredentials.setValue(
    'datasource',
    JSON.stringify({ type: datasourceObj.type, path: datasourceObj.path })
  )
  return Promise.all([
    sourceCredentials.toSecureString(masterPassword),
    passwordCredentials.toSecureString(masterPassword)
  ])
}

export const credentialsFromSecureStrings = async ({
  sourceCredentials,
  archiveCredentials,
  masterPassword
}) =>
  Promise.all([
    Credentials.fromSecureString(sourceCredentials, masterPassword),
    Credentials.fromSecureString(archiveCredentials, masterPassword)
  ])

export const credentialsToSecureStrings = async ({ sourceCredentials, archiveCredentials }) =>
  Promise.all([
    sourceCredentials.toSecureString(archiveCredentials.password),
    archiveCredentials.toSecureString(archiveCredentials.password)
  ])

export default {
  credentialsToWorkspace,
  passwordToSecureStrings,
  credentialsFromSecureStrings,
  credentialsToSecureStrings
}
