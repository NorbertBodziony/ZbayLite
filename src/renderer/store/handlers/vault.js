import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'

import { typeFulfilled, typeRejected, typePending, errorNotification } from './utils'
import nodeSelectors from '../selectors/node'
import identityHandlers from './identity'
import vaultHandlers from './vault'
import notificationsHandlers from './notifications'
import appHandlers from './app'

import vault from '../../vault'

export const VaultState = Immutable.Record({
  exists: false,
  creating: false,
  unlocking: false,
  creatingIdentity: false,
  locked: true,
  error: ''
}, 'VaultState')

export const initialState = VaultState()

export const actionTypes = {
  CREATE: 'CREATE_VAULT',
  UPDATE_IDENTITY_SIGNER_KEYS: 'UPDATE_IDENTITY_SIGNER_KEYS',
  SET_STATUS: 'SET_VAULT_STATUS',
  UNLOCK: 'UNLOCK_VAULT',
  CREATE_IDENTITY: 'CREATE_VAULT_IDENTITY',
  CLEAR_ERROR: 'CLEAR_VAULT_ERROR'
}

const createVault = createAction(actionTypes.CREATE, vault.create)
const unlockVault = createAction(
  actionTypes.UNLOCK,
  vault.unlock,
  ({ ignoreError = false }) => ({ ignoreError })
)
const createIdentity = createAction(actionTypes.CREATE_IDENTITY, vault.identity.createIdentity)
const updateIdentitySignerKeys = createAction(actionTypes.UPDATE_IDENTITY_SIGNER_KEYS, vault.identity.updateIdentitySignerKeys)
const clearError = createAction(actionTypes.CLEAR_ERROR)
const setVaultStatus = createAction(actionTypes.SET_STATUS)

export const actions = {
  createIdentity,
  updateIdentitySignerKeys,
  createVault,
  unlockVault,
  setVaultStatus,
  clearError
}

const loadVaultStatus = () => (dispatch, getState) => {
  const network = nodeSelectors.network(getState())
  return dispatch(setVaultStatus(vault.exists(network)))
}

const createVaultEpic = ({ name, password }, formActions) => async (dispatch, getState) => {
  const network = nodeSelectors.network(getState())
  try {
    await dispatch(appHandlers.actions.setNewUser(true))
    await dispatch(createVault({ masterPassword: password, network }))
    await dispatch(actions.unlockVault({
      masterPassword: password,
      createSource: true,
      network
    }))
    const identity = await dispatch(identityHandlers.epics.createIdentity({ name }))
    await dispatch(identityHandlers.epics.setIdentity(identity))
    await dispatch(setVaultStatus(true))
  } catch (error) {
    dispatch(notificationsHandlers.actions.enqueueSnackbar(
      errorNotification({ message: `Failed to create vault: ${error.message}` })
    ))
  }
  formActions.setSubmitting(false)
}

const unlockVaultEpic = ({ password: masterPassword }, formActions) => async (dispatch, getState) => {
  const state = getState()
  const network = nodeSelectors.network(state)
  try {
    await dispatch(vaultHandlers.actions.unlockVault({ masterPassword, network, ignoreError: true }))
    const [identity] = await vault.identity.listIdentities()
    await dispatch(identityHandlers.epics.setIdentity(identity))
  } catch (error) {
    if (error.message.includes('Authentication failed')) {
      dispatch(notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({ message: 'Incorrect password' }, { persist: false })
      ))
    } else {
      throw error
    }
  }
  formActions.setSubmitting(false)
}

export const epics = {
  loadVaultStatus,
  createVault: createVaultEpic,
  unlockVault: unlockVaultEpic
}

export const reducer = handleActions({
  [typePending(actionTypes.CREATE)]: state => state.set('creating', true),
  [typeFulfilled(actionTypes.CREATE)]: state => state.merge({
    creating: false,
    exists: true
  }),
  [typeRejected(actionTypes.CREATE)]: (state, { payload: error }) => state.merge({
    creating: false,
    error: error.message
  }),

  [typePending(actionTypes.UNLOCK)]: state => state.set('unlocking', true),
  [typeFulfilled(actionTypes.UNLOCK)]: (state, payload) => state.merge({
    unlocking: false,
    locked: false
  }),
  [typeRejected(actionTypes.UNLOCK)]: (state, { payload: error }) => state.merge({
    unlocking: false,
    locked: true,
    error: error.message
  }),

  [typePending(actionTypes.CREATE_IDENTITY)]: state => state.set('creatingIdentity', true),
  [typeFulfilled(actionTypes.CREATE_IDENTITY)]: state => state.set('creatingIdentity', false),
  [typeRejected(actionTypes.CREATE_IDENTITY)]: (state, { payload: error }) => state.merge({
    creatingIdentity: false,
    error: error.message
  }),
  [setVaultStatus]: (state, { payload: exists }) => state.set('exists', exists),
  [clearError]: state => state.delete('error')
}, initialState)

export default {
  actions,
  epics,
  reducer
}
