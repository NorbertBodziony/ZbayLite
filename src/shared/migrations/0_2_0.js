import * as Yup from 'yup'

import { getClient } from '../../renderer/zcash'
import { updateIdentity } from '../../renderer/vault'

const identityKeysSchema = Yup.object().shape({
  sk: Yup.string().required(),
  tpk: Yup.string().required()
})

const _getIdentityKeys = async (identity) => {
  try {
    return await Promise.all([
      await getClient().keys.exportTPK(identity.transparentAddress),
      await getClient().keys.exportSK(identity.address)
    ])
  } catch (err) {
    throw Error('Cannot migrate wallet to version 0.2.x - node doesn\'t contain identity keys.')
  }
}

const ensureIdentityHasKeys = async (identity) => {
  const haveKeys = await identityKeysSchema.isValid(identity.keys)
  if (!haveKeys) {
    const [tpk, sk] = await _getIdentityKeys(identity)
    const identityWithKeys = {
      ...identity,
      keys: {
        tpk,
        sk
      }
    }
    await updateIdentity(identityWithKeys)
    return identityWithKeys
  }
  return identity
}

export default {
  ensureIdentityHasKeys
}
