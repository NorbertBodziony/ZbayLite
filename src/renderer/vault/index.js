import fs from 'fs'
import path from 'path'
import * as Yup from 'yup'
import { DateTime } from 'luxon'

import { remote } from 'electron'

import { passwordToSecureStrings } from './marshalling'
import { formSchema as shippingDataSchema } from '../components/widgets/settings/ShippingSettingsForm'
import Vault from './vault'

const getVaultPath = network => path.join(remote.app.getPath('userData'), `vault-${network}.bcup`)

let _vault = null

export const withVaultInitialized = fn => (...args) => {
  if (!_vault) {
    throw Error('Archive not initialized.')
  }
  return fn(...args)
}

export const exists = network => fs.existsSync(getVaultPath(network))

export const create = async ({ masterPassword, network }) => {
  if (_vault !== null) {
    throw Error('Archive already initialized.')
  }
  const datasourceObj = {
    type: 'ipc',
    path: getVaultPath(network)
  }
  const [sourceCredentialsStr, passwordCredentialsStr] = await passwordToSecureStrings({
    masterPassword,
    datasourceObj
  })
  _vault = new Vault(sourceCredentialsStr, passwordCredentialsStr)
}

export const unlock = async ({ masterPassword, network, createSource = false }) => {
  if (!_vault) {
    await create({ masterPassword, network })
  }
  try {
    await _vault.unlock(masterPassword, createSource)
  } catch (err) {
    _vault = null
    throw err
  }
}

export const lock = async () => _vault.lock()

const _entryToIdentity = entry => {
  const entryObj = entry.toObject()
  return {
    id: entryObj.id,
    name: entryObj.properties.name,
    address: entryObj.properties.address,
    signerPrivKey: entryObj.properties.signerPrivKey,
    signerPubKey: entryObj.properties.signerPubKey,
    transparentAddress: entryObj.properties.transparentAddress,
    keys: JSON.parse(entryObj.properties.keys || '{}'),
    shippingData: JSON.parse(entryObj.properties.shippingData || '{}'),
    donationAllow: entryObj.properties.donationAllow || 'true',
    shieldingTax: entryObj.properties.shieldingTax || 'true',
    donationAddress:
      entryObj.properties.donationAddress ||
      (parseInt(process.env.ZBAY_IS_TESTNET)
        ? 'ztestsapling1e2fr65z5t537cyahpu55fm335a2z26293nvj4wu0jjw55s66r669kaw4k26hmgnk23zeqw2k74n'
        : 'zs19wdfjhfqa6afgyv2vw07f68zmwhudgqwa5sve3zgz8y856nhn3j73yr9tavugfs8nunn6pju34a') // TODO add real donation address
  }
}

export const getVault = withVaultInitialized(() => _vault)

const newIdentitySchema = Yup.object().shape({
  name: Yup.string().required(),
  transparentAddress: Yup.string().required(),
  address: Yup.string().required(),
  signerPrivKey: Yup.string().required(),
  signerPubKey: Yup.string().required(),
  keys: Yup.object()
    .shape({
      sk: Yup.string().required(),
      tpk: Yup.string().required()
    })
    .required()
})

export const createIdentity = async identity => {
  let entry = null
  await newIdentitySchema.validate(identity)
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup
      .createEntry(identity.name)
      .setProperty('name', identity.name)
      .setProperty('address', identity.address)
      .setProperty('transparentAddress', identity.transparentAddress)
      .setProperty('signerPubKey', identity.signerPubKey)
      .setProperty('signerPrivKey', identity.signerPrivKey)
      .setProperty('keys', JSON.stringify(identity.keys))
    workspace.save()
  })
  return _entryToIdentity(entry)
}

const updateIdentitySchema = Yup.object().shape({
  id: Yup.string().required(),
  name: Yup.string().required(),
  transparentAddress: Yup.string().required(),
  address: Yup.string().required(),
  keys: Yup.object()
    .shape({
      sk: Yup.string().required(),
      tpk: Yup.string().required()
    })
    .required()
})

export const updateIdentity = async identity => {
  let entry = null
  await updateIdentitySchema.validate(identity)
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup
      .findEntryByID(identity.id)
      .setProperty('name', identity.name)
      .setProperty('address', identity.address)
      .setProperty('transparentAddress', identity.transparentAddress)
      .setProperty('keys', JSON.stringify(identity.keys))
    workspace.save()
  })
  return _entryToIdentity(entry)
}

export const updateShippingData = async (identityId, shippingData) => {
  let entry = null
  await shippingDataSchema.validate(shippingData)
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup
      .findEntryByID(identityId)
      .setProperty('shippingData', JSON.stringify(shippingData))
    workspace.save()
  })
  return _entryToIdentity(entry)
}

const updateIdentitySignerKeysSchema = Yup.object().shape({
  signerPubKey: Yup.string().required(),
  signerPrivKey: Yup.string().required()
})

export const updateIdentitySignerKeys = async identity => {
  let entry = null
  await updateIdentitySignerKeysSchema.validate(identity)
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup
      .findEntryByID(identity.id)
      .setProperty('signerPubKey', identity.signerPubKey)
      .setProperty('signerPrivKey', identity.signerPrivKey)
    workspace.save()
  })
  return _entryToIdentity(entry)
}

const updateDonationSchema = Yup.boolean().required()

export const updateDonation = async (identityId, allow) => {
  let entry = null
  await updateDonationSchema.validate(allow)
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup.findEntryByID(identityId).setProperty('donationAllow', allow.toString())
    workspace.save()
  })
  return _entryToIdentity(entry)
}
export const updateShieldingTax = async (identityId, allow) => {
  let entry = null
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup.findEntryByID(identityId).setProperty('shieldingTax', allow.toString())
    workspace.save()
  })
  return _entryToIdentity(entry)
}

export const updateDonationAddress = async (identityId, address) => {
  let entry = null
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup
      .findEntryByID(identityId)
      .setProperty('donationAddress', address.toString())
    workspace.save()
  })
  return _entryToIdentity(entry)
}
export const updateLastLogin = async identityId => {
  let entry = null
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup.findEntryByID(identityId).setProperty(
      'lastLogin',
      DateTime.utc()
        .toSeconds()
        .toString()
    )
    workspace.save()
  })
  return _entryToIdentity(entry)
}
export const getLastLogin = async identityId => {
  let entry = null
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup.findEntryByID(identityId).getProperty('lastLogin')
  })
  return entry
}

export const listIdentities = async () => {
  let identities = []
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    identities = identitiesGroup.getEntries()
  })
  return identities.map(_entryToIdentity)
}

export const remove = async () => {
  if (_vault) {
    await lock()
    _vault = null
  }
}

export const locked = () => {
  return _vault === null || _vault.locked()
}

export default {
  // TODO: [refactoring] those using withWorkspace should be moved to Vault
  // as plug in modules
  identity: {
    createIdentity: withVaultInitialized(createIdentity),
    getLastLogin: withVaultInitialized(getLastLogin),
    updateLastLogin: withVaultInitialized(updateLastLogin),
    listIdentities: withVaultInitialized(listIdentities),
    updateShippingData: withVaultInitialized(updateShippingData),
    updateIdentitySignerKeys: withVaultInitialized(updateIdentitySignerKeys),
    updateDonation: withVaultInitialized(updateDonation),
    updateDonationAddress: withVaultInitialized(updateDonationAddress),
    updateShieldingTax: withVaultInitialized(updateShieldingTax)
  },
  locked,
  getVault,
  remove,
  exists,
  create,
  unlock,
  lock: withVaultInitialized(lock)
}
