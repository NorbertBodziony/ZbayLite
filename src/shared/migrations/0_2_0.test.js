/* eslint import/first: 0 */
jest.mock('../../renderer/zcash')
jest.mock('../../renderer/vault/vault')
jest.mock('../../renderer/vault/marshalling', () => ({
  passwordToSecureStrings: jest.fn(() => [jest.mock(), jest.mock()]),
  createArchive: jest.requireActual('../../renderer/vault/marshalling').createArchive
}))

import { create, remove, getVault, listIdentities } from '../../renderer/vault'
import { mock } from '../../renderer/vault/vault'
import { createArchive } from '../../renderer/vault/marshalling'
import { mock as zcashMock } from '../../renderer/zcash'
import * as R from 'ramda'

import migrations from './0_2_0'

describe('0.2.0 migrations', () => {
  beforeEach(async () => {
    jest.clearAllMocks()
  })

  describe('- ensureIdentityHasKeys', () => {
    const sk = 'sapling-private-address-spending-key'
    const tpk = 'transparent-address-private-key'
    const identity = {
      name: 'Saturn',
      address: 'sapling-private-address',
      transparentAddress: 'transparent-address'
    }

    const identityWithKeys = {
      ...identity,
      keys: {
        sk, tpk
      }
    }

    beforeEach(async () => {
      await remove()
      await create({ masterPassword: 'test password' })
      mock.workspace.archive = createArchive()
      mock.workspace.save = jest.fn()
    })

    it('does not run if identity has keys', async () => {
      await migrations.ensureIdentityHasKeys(identityWithKeys)

      expect(mock.workspace.save).not.toHaveBeenCalled()
      expect(zcashMock.requestManager.z_exportkey).not.toHaveBeenCalled()
      expect(zcashMock.requestManager.dumpprivkey).not.toHaveBeenCalled()
    })

    it('updates identity in the vault with keys', async () => {
      let entry
      await getVault().withWorkspace(workspace => {
        const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
        entry = identitiesGroup.createEntry(identity.name)
          .setProperty('name', identity.name)
          .setProperty('address', identity.address)
          .setProperty('transparentAddress', identity.transparentAddress)
      })
      const identityFromVault = {
        ...identity, id: entry.id
      }

      await migrations.ensureIdentityHasKeys(identityFromVault)

      expect(zcashMock.requestManager.z_exportkey).toHaveBeenCalledWith(identity.address)
      expect(zcashMock.requestManager.dumpprivkey).toHaveBeenCalledWith(identity.transparentAddress)
      expect(mock.workspace.save).toHaveBeenCalled()
      const identities = await listIdentities()
      expect(identities.map(R.omit(['id']))).toMatchSnapshot()
    })

    it('throws if not migrated on a new node', async () => {
      let entry
      await getVault().withWorkspace(workspace => {
        const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
        entry = identitiesGroup.createEntry(identity.name)
          .setProperty('name', identity.name)
          .setProperty('address', identity.address)
          .setProperty('transparentAddress', identity.transparentAddress)
      })
      const identityFromVault = {
        ...identity, id: entry.id
      }
      zcashMock.requestManager.z_exportkey.mockRejectedValueOnce(
        new Error('this address doesn\t belong to the wallet')
      )
      zcashMock.requestManager.dumpprivkey.mockRejectedValueOnce(
        new Error('this address doesn\t belong to the wallet')
      )

      expect.assertions(2)

      try {
        await migrations.ensureIdentityHasKeys(identityFromVault)
      } catch (err) {
        expect(err).toMatchSnapshot()
      }

      expect(mock.workspace.save).not.toHaveBeenCalled()
    })
  })
})
