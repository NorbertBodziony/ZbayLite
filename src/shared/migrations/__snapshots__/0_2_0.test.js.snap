// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0.2.0 migrations - ensureIdentityHasKeys throws if not migrated on a new node 1`] = `[Error: Cannot migrate wallet to version 0.2.x - node doesn't contain identity keys.]`;

exports[`0.2.0 migrations - ensureIdentityHasKeys updates identity in the vault with keys 1`] = `
Array [
  Object {
    "address": "sapling-private-address",
    "keys": Object {
      "sk": "sapling-private-address-spending-key",
      "tpk": "transparent-address-private-key",
    },
    "name": "Saturn",
    "signerPrivKey": undefined,
    "signerPubKey": undefined,
    "transparentAddress": "transparent-address",
  },
]
`;
